File: TUnit.TestProject.FSharp/ArgumentWithImplicitConverterTests.fs
namespace TUnit.TestProject.FSharp

open System
open TUnit.Core

// F# equivalents for ArgumentWithImplicitConverterTests.cs

type ExplicitInteger =
    | ExplicitInteger of int
    static member op_Explicit(i: int) = ExplicitInteger i
    override this.ToString() =
        let (ExplicitInteger i) = this in i.ToString()

type ImplicitInteger =
    | ImplicitInteger of int
    static member op_Implicit(i: int) = ImplicitInteger i
    override this.ToString() =
        let (ImplicitInteger i) = this in i.ToString()

type ArgumentWithImplicitConverterTests() =
    [<Test>]
    [<Arguments(1)>]
    [<Arguments(2)>]
    [<Arguments(3)>]
    member _.Explicit(integer: ExplicitInteger) =
        Console.WriteLine(integer)

    [<Test>]
    [<Arguments(1)>]
    [<Arguments(2)>]
    [<Arguments(3)>]
    member _.Implicit(integer: ImplicitInteger) =
        Console.WriteLine(integer)

File: TUnit.TestProject/ArgumentWithImplicitConverterTests.cs
using TUnit.TestProject.Attributes;

namespace TUnit.TestProject;

[EngineTest(ExpectedResult.Pass)]
public class ArgumentWithImplicitConverterTests
{
    [Test]
    [Arguments(1)]
    [Arguments(2)]
    [Arguments(3)]
    public void Explicit(ExplicitInteger integer)
    {
        Console.WriteLine(integer);
    }

    [Test]
    [Arguments(1)]
    [Arguments(2)]
    [Arguments(3)]
    public void Implicit(ImplicitInteger integer)
    {
        Console.WriteLine(integer);
    }
}

public readonly struct ExplicitInteger(int i)
{
    public static explicit operator ExplicitInteger(int i) => new(i);

    public override string ToString()
    {
        return i.ToString();
    }
}


public readonly struct ImplicitInteger(int i)
{
    public static implicit operator ImplicitInteger(int i) => new(i);

    public override string ToString()
    {
        return i.ToString();
    }
}
